### ISSUE TITLE
Can INVPROP prove the property at intermediate layers?

### ISSUE THREAD

**Issue by ytsao** [kothasuhas/verify-input#6]
Hi,

It seems to me, INVPROP is taking output constraints (property) into account for tightening bounds for each neuron.
If we would like to prove the property, we still need to do forward pass again to see if the property can be proved by new output bounds.

I'm wondering **if INVPROP can prove the property at intermediate layers?** like forward-backward abstract interpretation (Scalable verification of GNN-based job schedulers).

If my understanding about INVPROP is not correct, please let me know. 
Thank you very much.

Yi-Nung

**Comment by ChristopherBrix** [comment]
The output constraints are included in the optimization of all affected neurons via Lagrangian multipliers. So at an intermediate layer, the bounds are optimized at the same time with respect to the input bounds, potential branching decisions, and the output constraints. If the problem is UNSAT (no adversarial examples exist), the output constraints are unsatisfiable. This will cause the lower bounds to eventually exceed the upper bounds (mathematically, we're optimizing over an empty set, so we'd get a lower bound of plus infinity). At that point, we can stop the optimization and return UNSAT.

Does that answer your questions? If not, please let me know!

**Comment by ytsao** [comment]
Thanks for the swift reply.

So, basically, INVPROP is trying to solve the max-min optimization problem which is shown below Algorithm 1 in the paper.
But, I still cannot understand the part that **the bounds are optimized**. 
Is that meaning the tightened bounds will be released after solving this optimization problem? 
It seems that we still have to solve two optimization problems for each neuron to refine its lower and upper bounds, right?
If so, does INVPROP really solve 2n optimization problems or just partial? (n is the number of neurons in the network.)

Going back to your explanation about UNSAT, in INVPROP, basically, your verification decision is upon if there exists lower bound exceed upper bound or not. If so, return UNSAT, otherwise, we still have tightened bounds for each neuron such that the effort from branch-and-bound can be reduced (since the search space is smaller).

If I misunderstood some parts, please let me know. 
Thanks

Yi-Nung

**Comment by ChristopherBrix** [comment]
I'm not quite sure what you mean with "released". After the optimization you can indeed access the bounds of all intermediate layers. Those will include the benefit of being tightened by the output constraints, so they'll (hopefully) be tighter than what you'd get without INVPROP.

You could implement all of this as solving 2n problems. The same is true for abCROWN without INVPROP: You could optimize each bound individually. However, abCROWN uses a trick: The bounds of intermediate neurons are computed symbolically, and then immediately used to compute the bounds for the next layer. Thus, each layers bounds depend on the bounds of all previous layers. That's why only the lower bound of the final output node needs to be optimized: That lower bound will improve if the lower/upper bounds of intermediate neurons will improve, so they're optimized implicitly.
So we only solve one optimization problem, but implicitly, the bounds of all neurons are optimized (with the exception of bounds that don't actually impact the final output bound, e.g. if something would be multiplied by a weight of 0).

And yes, if INVPROP fails to prove UNSAT, then the tightened bounds will be used during the branch-and-bound process.

**Comment by ytsao** [comment]
Thanks for the reply.

Ok, so, after solving this optimization problem (either find the optimal or meet the termination condition), actually, we will have all tighter bounds (includes the output neurons) such that we could check if the updated output bounds are able to prove UNSAT, is it right?
Also, for this single optimization problem, is it shown in Appendix D in the paper?

For the implementation, I could get the concept of this trick used in abcrown.
Could I also ask where is the implementation about that in abcrown repository? Thanks.

**Comment by ChristopherBrix** [comment]
Yes, you could get the updated bounds after they've been tightened by aCROWN and INVPROP. In practice, if all you care about is knowing UNSAT/SAT, then abCROWN will just return that by default.
If you do extract the bounds, they may be
a) Stronger if you used INVPROP than without it, because the output constraints were beneficial
b) The same as without INVPROP, because they didn't help
c) Weaker, because unfortunately the bound tightening process is a complex optimization problem and may find local minima, so INVPROP could potentially cause abCROWN to converge to a different (and worse) local minimum.

Yes, Appendix D should formally describe that optimization process.

The trick of only optimizing one problem is implemented implicitly: Assume you have a simple problem
`min_x f(x) = g(x) + h(x)`. Where `x in [2,3]`, `g(x) = alpha_g * x` and `h(x) = alpha_h * x` with `alpha_g, alpha_h in [0, 1]` and initially set to 0.5. Crucially, the `alpha` are implemented as PyTorch variables. So when we compute the lower bound on `g`, we get `OptimizableVariable[alpha_g] * x` at the input level. We know that `alpha_g >= 0`, so we assign the lower bound of `x` (if `alpha_g` were negative, we'd need the upper bound of `x` to compute the lower bound of the formula). So this becomes `lower_bound(g) = 2 * OptimizableVariable[alpha_g]`.
Now, we can infer a concrete value, by inserting the current concrete value of `alpha_g = 0.5`. However, instead, when we want to compute the lower bound of `f` and need to access the lower bound of `g`, we use the representation of `OptimizableVariable[alpha_g]`. So now, the bound of `f` depends on the optimized value of `alpha_g`, and gradient descent will optimize it for us.

So there's not an obvious point in the code where we have a toggle that switches this trick on or off. It happens every time we store a lower/upper bound as a formula dependent on its optimizable variables.

**Comment by ytsao** [comment]
OK, thanks for the reply.
I will check how to get the bounds from abcrown later. 
Because I wasn't able to find the function or code to get bounds before.

I see the process, thanks. 
Further small question, in alpha-crown, also need to solve an optimization problem to optimize `alpha` to have better lower bound. So, when will INVPROP be applied? Is it after the process of alpha-crown, or both methods will be used at the same time if the `alpha` is the same?

Thanks for all explanation, I'm getting clear for this work.

**Comment by ChristopherBrix** [comment]
a-CROWN and INVPROP are applied in parallel.

If you want to access the bounds after the a-CROWN (and INVPROP) phase, but before b-CROWN (which will then not be run), please have a look at https://github.com/kothasuhas/verify-input/blob/873d80ac148b27be5b52a381f7d4fe63ae268616/examples/double_integrator/files/oc.py#L49-L54

**Comment by ytsao** [comment]
So, INVPROP doesn't need the bounds from a-CROWN.  
What would be the initial bounds for INVPROP then? Is it like the iteratively tighten by each?
Initially, I thought INVPROP would be executed after a-CROWN, since it requires the bounds from cheap forward method which is mentioned in Algorithm 1 in the paper.

Thanks for the code sharing!

**Comment by ChristopherBrix** [comment]
The bounds are initialized via IBP. They are then iteratively tightened. That optimization process involves both a-CROWN and INVPROP.

**Comment by ytsao** [comment]
Ok, I see. Thank you sooo much :)
